src/main/java/com/code_intelligence/jazzer/driver/OfflineInstrumentor.java

```diff
@@ -0,0 +1,159 @@
+// Copyright 2021 Code Intelligence GmbH
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.code_intelligence.jazzer.driver;
+
+import com.code_intelligence.jazzer.agent.AgentInstaller;
+import com.code_intelligence.jazzer.utils.Log;
+import com.code_intelligence.jazzer.utils.ZipUtils;
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.lang.UnsupportedClassVersionError;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.net.URLClassLoader;
+import java.util.ArrayList;
+import java.util.Enumeration;
+import java.util.HashSet;
+import java.util.List;
+import java.util.jar.JarEntry;
+import java.util.jar.JarFile;
+import java.util.zip.ZipOutputStream;
+
+public class OfflineInstrumentor {
+  /**
+   * Create a new jar with classes that have Jazzer instrumentation
+   *
+   * @param jarLists list of jars to instrument
+   * @return a list of errors that were hit when trying to instrument all classes in jar
+   */
+  public static boolean instrumentJars(List<String> jarLists) {
+    if (jarLists.size() > 1) {
+      Log.error("Only instrumenting one jar at a time is currently supported");
+      return false;
+    }
+
+    AgentInstaller.install(Opt.hooks);
+
+    List<String> errorMessages = new ArrayList<String>();
+    for (String jarPath : jarLists) {
+      try {
+        Log.info("Instrumenting jar file: " + jarPath);
+        errorMessages = createInstrumentedClasses(jarPath);
```

Comment:
I think the resulting jar file could still be useful. This will mean that instrumenting one or more classes failed. If it is one or two classes out of **n** classes, and n-2 classes are still instrumented, then the result could still be good, and the failed classes will still be in your output jar, just not instrumented.

This path we will still fail and return exit(1), but we will also create the instrumented jar before we fail so if the classes that weren't instrumented are OK to skip, the user will still have their instrumented jar.

Reply:
Sounds reasonable.